woocommerce/google-listings-and-ads
# Save edited free campaign.

### Changes proposed in this Pull Request:

- Make `EditFreeCampaign` component manage the state of the form from `ChooseAudience` and `SetupFreeListings`.
- Preserve the state when navigating between steps.
- Save basic config: target audience and settings, then sync.
- Redirect to the dashboard and record a tracking event once saved & synced.
- Open the edit page in the same tab as the dashboard.

Implements part of https://github.com/woocommerce/google-listings-and-ads/issues/156

### Follow-ups:
This PR is mostly, about moving state management and saving functionality to the "main" form, so there is a list of other things to be implemented separately not to blur the picture.

- [x] Prompt on leave  - Work was already done at https://github.com/woocommerce/google-listings-and-ads/compare/feature/156-save-edited-campaign...feature/156-promptEffect?expand=1. I created a separate https://github.com/woocommerce/google-listings-and-ads/pull/393 to make it easier to review and merge. We will need one more commit to integrate those.
- [ ] Save Shipping rates and times.
- [ ] Notify errors for individual save requests.
- [ ] Update the countries list on step 2 when changed in step 1.
- [ ] Disable the form and submit button while saving.
- [ ] Highlighting edited step on the dashboard page. The `programId` is being forwarded from this form.

### Screenshots:


https://user-images.githubusercontent.com/17435/112766335-a3177780-9011-11eb-8d44-c9bf865d8ecc.mp4

(Please note, that on the second step only settings managed by `/mc/settings` are saved: radio buttons, checkboxes and free shipping threshold. Per-country rate and time is **not** covered by this PR)


### Detailed test instructions:

1. Go to **Marketing > GLA > Edit "  https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fedit-free-campaign&programId=123&pageStep=2
2. Do some changes in the first step
3. Navigate forward (with continue button, or step title)
4. Do some changes to the second step
5. (optionally) navigate back and forth
6. Click final "Save Changes"
7. Check if the look matches Figma designs
8. Check if the data is saved after clicking the final "save" and not before.

### Changelog Note:

> Add editing free campaign feature.


## Doubts:
1. `gla_free_campaign_edited` tracking event does not send any additional data, should it? //cc @jconroy 
2. We send a number (currently 3, will be more) of saving requests to the server-side. Do we want some specific way to notify the partial success & failure? When, for example, due to flaky internet connection Target Audience was saved successfully, but Shipping not, or saving to local DB succeeded, but syncing with Google didn't. //cc @j111q
